Use clamped position value for find*WordBoundary calls

fantasai opened the pull request:
#### c0a0c9ebd27a381e249128864ed5f3d505494dc4
<pre>
Use clamped position value for find*WordBoundary calls
<a href="https://bugs.webkit.org/show_bug.cgi?id=276002">https://bugs.webkit.org/show_bug.cgi?id=276002</a>
<a href="https://rdar.apple.com/130331432">rdar://130331432</a>

Reviewed by David Kilzer and Ryosuke Niwa.

Using the unclamped value is nonsense.

* Source/WebCore/platform/text/mac/TextBoundaries.mm:
(WebCore::findWordBoundary):

Canonical link: <a href="https://commits.webkit.org/280503@main">https://commits.webkit.org/280503@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ae687538c5f5f37c86a12f6ffdaabe697d7f3ff9

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/56758 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/36086 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/9232 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/60373 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/7201 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/58885 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/43709 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/7391 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/45979 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/5054 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/58788 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/33915 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/48990 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/26836 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/30695 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/6321 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/6204 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/52673 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/6592 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/62056 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/671 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/6693 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/53240 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/674 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/49045 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/53258 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12563 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/578 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/31915 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/33000 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/34085 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/32746 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

ddkilzer commented on Source/WebCore/editing/TextIterator.cpp:
```diff
@@ -2215,6 +2215,7 @@ inline bool SearchBuffer::isWordEndMatch(size_t start, size_t length) const
 {
     ASSERT(length);
     ASSERT(m_options.contains(FindOption::AtWordEnds));
+    ASSERT(start + length <= m_buffer.size());
```
I think we should make this `ASSERT_WITH_SECURITY_IMPLICATION()` since it has security implications.

However, I think this is the same check done on Line 2329 (now 2330) in `TextIterator.cpp` before calling this method, so it's probably redundant and can be removed.

If you keep it, please switch to ASSERT_WITH_SECURITY_IMPLICATION()`.

```
    ASSERT_WITH_SECURITY_IMPLICATION(matchStart + matchedLength <= size);
```

ddkilzer commented on Source/WebCore/platform/text/mac/TextBoundaries.mm:
```diff
@@ -191,32 +233,47 @@ void findWordBoundary(StringView text, int position, int* start, int* end)
-        pos = text.length() - 1;
+    if (static_cast<unsigned>(position) >= text.length()) {
+        // This makes no sense. Return what we got in.
+        ASSERT(true);
```
Let's re-use the failed check (reverse the logic) and document the bug with `ASSERT_WITH_MESSAGE()` rather than a comment (please feel free to change the comment, or omit it if the expression is sufficient):

```suggestion
        ASSERT_WITH_MESSAGE(static_cast<unsigned>(position) < text.length(), "position exceeds text.length()");
```

ddkilzer commented on Source/WebCore/platform/text/mac/TextBoundaries.mm:
```diff
@@ -191,32 +233,47 @@ void findWordBoundary(StringView text, int position, int* start, int* end)
+        // This makes no sense. Return what we got in.
+        ASSERT(true);
+        *start = position;
+        *end = position;
```
I don't think we should set `*start` and `*end` to values outside the range as that could cause a crash later.

I think we should clamp the values to `text.length() - 1`.

```suggestion
        *start = text.length() - 1;
        *end = text.length() - 1;
```

rniwa commented on Source/WebCore/platform/text/mac/TextBoundaries.mm:
```diff
@@ -91,16 +91,24 @@ static CFStringTokenizerRef tokenizerForString(CFStringRef str)
     ASSERT(position >= 0);
     ASSERT(static_cast<unsigned>(position) < text.length());
 
+    auto string = (is8Bit) ? text.span8() : text;
```
This parenthesis seems unnecessary?

fantasai replied:
I always forget about that for some reason. Will fix.

rniwa commented on Source/WebCore/platform/text/mac/TextBoundaries.mm:
```diff
@@ -91,16 +91,24 @@ static CFStringTokenizerRef tokenizerForString(CFStringRef str)
 }
 
 // Simple case: A word is a stream of characters delimited by a special set of word-delimiting characters.
+template<bool is8Bit = false>
```
Instead of boolean, can we define an enum class for this?
e.g. `CharType : bool { Is8Bit, Is16Bit };`

fantasai replied:
I did that originally, since it makes the call site more understandable; but it felt a lot less readable within the function, where it is very intrusive. This at least mirrors the StringView's own is8Bit interface. Also, since this file is all public and not namespaced, it injects CharType into the wider world; idk if we want that.

rniwa replied:
ok, sounds reasonable to me.

aprotyas commented on Source/WebCore/platform/text/mac/TextBoundaries.mm:
```diff
@@ -191,32 +233,47 @@ void findWordBoundary(StringView text, int position, int* start, int* end)
+            findSimpleWordBoundary<true>(text, position, start, end);
+        else
+            findSimpleWordBoundary<false>(text, position, start, end);
+    }
```
I think you can reuse the template arg
```suggestion
    else
        findSimpleWordBoundary<is8Bit>(text, position, start, end);
```

fantasai replied:
It's not a template arg in this function though, so it has to be a literal. :)

aprotyas replied:
~~I thinkkkk this would still work, but again, just a nit anyway!~~ Ignore, of course it wouldn't work.

fantasai replied:
I get it. I make the same mistake every time I have to use templates, and the compiler always tells me off. :p

megangardner commented on Source/WebCore/platform/text/mac/TextBoundaries.mm:
```diff
@@ -191,32 +233,47 @@ void findWordBoundary(StringView text, int position, int* start, int* end)
+        *start = position;
+        *end = position;
+        return;
+    }
```
I'm not sure I follow the above. Instead of returning the end of the text if the position is beyond it, you think we should just return the original position? I think both are valid, but I worry that places might be depending on this behavior, but I haven't investigated. I'm also not sure why you assert(true)?

fantasai replied:
I'm asserting because calling this function with a position beyond the length of the string is, afaict, complete nonsense (and would have crashed before this change). But in order to avoid a crash, we need to return something so rather than asserting the inequality twice I included ASSERT(true) in the `if` statement.

ddkilzer reviewed (approved):
A couple minor requests, but overall looks good!

fantasai committed:
Use clamped position value for find*WordBoundary calls
https://bugs.webkit.org/show_bug.cgi?id=276002
rdar://130331432

Reviewed by David Kilzer and Ryosuke Niwa.

Using the unclamped value is nonsense.

* Source/WebCore/platform/text/mac/TextBoundaries.mm:
(WebCore::findWordBoundary):

Canonical link: https://commits.webkit.org/280503@main
